### PR TITLE
 Set type of symlink parameter `type` to a union of `"file" | "dir" | "junction" | undefined | null`

### PR DESCRIPTION
**Pull request by KSXGitHub**
### Additional changes

* Add an overload to `fsExtra.symlink`

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by shiftkey** [comment]
```
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/fs-extra-promise/fs-extra-promise-tests.ts:121:34
ERROR: 121:34  expect  TypeScript@next compile error: 
Argument of type 'string' is not assignable to parameter of type 'Type'.
    at /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:101:19
    at Generator.next (<anonymous>)
    at fulfilled (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:5:58)
    at <anonymous>
```

@KSXGitHub looks like this test needs to be reviewed:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/683657958096b2cf47f98cdf77aedabb5d84c35e/types/fs-extra-promise/fs-extra-promise-tests.ts#L121

**Comment by KSXGitHub** [comment]
@shiftkey I just pushed a fix, hopefully no more packages appear.

**Comment by KSXGitHub** [comment]
For some reason, my pull requests always fail at CI first run, probably because of memory bottleneck? Someone should fix the CI.

**Review by alan-agius4 — APPROVED** [review]

**Review by BrunoScheufler — APPROVED** [review]
Looks great 👍

**Review by jrockwood — APPROVED** [review]
Yep, looks good. Thanks for making the change!

File: .gitignore [hunk 1]
```diff
@@ -49,3 +49,6 @@ npm-debug.log
 
 # yarn
 yarn.lock
+
+# pnpm
+shrinkwrap.yaml
```
Comment (uniqueiniquity) [inline .gitignore]: Please do not include this change to the `.gitignore`. Thanks!

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (KSXGitHub) [reply .gitignore]: Why shouldn't I?

Comment (RyanCavanaugh) [reply .gitignore]: Please send non-types (i.e. infrastructure) changes in a separate PR along with reasoning for why the change is needed. Mixing/matching different changes makes it harder to smartly roll back changes if needed.

Comment (KSXGitHub) [reply .gitignore]: OK, I just moved it to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26192

**Review by mohsen1 — APPROVED** [review]

**Review by uniqueiniquity — APPROVED** [review]

**Comment by jdneo** [comment]
Got this error after using the new version: 5.0.3

```
node_modules/@types/fs-extra/index.d.ts(210,86): error TS2694: Namespace '"fs"' has no exported member 'symlink'.
node_modules/@types/fs-extra/index.d.ts(212,87): error TS2694: Namespace '"fs"' has no exported member 'symlink'.
```

node version: 8.11.2

**Comment by KSXGitHub** [comment]
@jdneo It seems that your `@types/fs-extra` uses older version of `@types/node`. You should try update `@types/node`.

**Comment by jeremejevs** [comment]
@KSXGitHub I'm building an app for Electron 2, which is based on Node 8.9.3, so my `@types/node` is pinned at `~8.9`, and the typings for that version [have no `Type` in the `symlink` namespace](https://github.com/DefinitelyTyped/DefinitelyTyped/blame/9f4c75126167d0d8af759f58405d53d983e94ad0/types/node/v8/index.d.ts#L3368-L3377). Not sure how to solve this properly, besides using a specific version of `@types/node` in this package. Can't make a PR for `@types/node@8.9`, since it's now permanently archived.

**Comment by KSXGitHub** [comment]
> Can't make a PR for @types/node@8.9, since it's now permanently archived.

I made a [change](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25975/files#diff-8ed6e12981d3914d0152a41c16fe2980) to v8 though.

> so my @types/node is pinned at `~8.9`

Why not `^8.9`?

**Comment by jeremejevs** [comment]
Because I'm using Node 8.9, so I don't want 8.10 or 8.11 typings, which may permit me to use some API which doesn't exist in 8.9.

**Comment by KSXGitHub** [comment]
@jdneo @jeremejevs I created https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26381

**Comment by jeremejevs** [comment]
Hmm, I don't think this needs to be solved ASAP, does it? I'm personally fine with staying on `@types/fs-extra@5.0.2` for some while. I'd see where the #26378 goes first, it would be a cleaner fix to just narrow `@types/node` in this repo to something like `^8.10 || ^9.foo || ^10.bar || >= 11`, if it's okay.

**Comment by Huachao** [comment]
@KSXGitHub any updates on this?

**Comment by KSXGitHub** [comment]
@Huachao I have already created a [pull request](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26381) (although it's likely not going to be accepted) so I worry about this no longer. In the meantime, you can either pin your `@types/fs-extra` version or just update `@types/node` already.

**Pull request merged**